curriculum/challenges/english/25-front-end-development/lab-book-organizer/67172b43f84bcd2dec238a3d.md

```diff
@@ -0,0 +1,164 @@
+
+8. You should call the `sort` higher order function by passing the `sortByYear` callback function on the `filteredBooks` array to sort the books according to their `releaseYear` in ascending order and save the sorted array in a new array named `filteredBooksSorted`.
+
+9. You should call the `forEach` higher order function on the `filteredBooksSorted` array and print out the details of each book following the `<BOOK_TITLE> by <BOOK_AUTHOR> (<releaseYear>)` format.
+
+# --hints--
+
+You should have a function `sortByYear` in your code.
+
+```js
+assert.isFunction(sortByYear);
+```
+
+Your `sortByYear` function should return `-1` if the `releaseYear` of `book1` object is smaller than that of the `book2` object, `1` if the `releaseYear` of `book1` object is larger than that of the `book2` object, and `0` in all other scenarios.
+
+```js
+assert.equal(sortByYear({releaseYear: 1913}, {releaseYear: 1925}), -1);
+assert.equal(sortByYear({releaseYear: 1925}, {releaseYear: 1913}), 1);
+assert.equal(sortByYear({releaseYear: 1925}, {releaseYear: 1925}), 0);
+```
+
+You should have an array `books` in your code.
+
+```js
+assert.isArray(books)
+```
+
+Your `books` array should have at least three book objects.
+
+```js
+assert.isAtLeast(books.length, 3)
+```
+
+Your `books` array should contain objects each with a string `title`, another string `authorName`, and a number `releaseYear`.
+
+```js
+
+books.forEach((book) => {
+  assert.isObject(book);
+  assert.sameMembers(Object.keys(book), ["title", "authorName", "releaseYear"]);
+  assert.isString(book.title);
+  assert.isNotEmpty(book.title);
+  assert.isString(book.authorName);
+  assert.isNotEmpty(book.authorName);
+  assert.isNumber(book.releaseYear);
+})
+```
+
+You should have an array `filteredBooks` in your code.
+
+```js
+assert.isArray(filteredBooks)
+```
+
+You should have an array `filteredBooksSorted` in your code.
+
+```js
+assert.isArray(filteredBooksSorted)
+```
+
+You should call the `filter` higher-order function on the `books` array with a callback to filter out books and save the filtered array in a new array named `filteredBooks`.
+
+```js
+assert.match(code, "(let|const)\s{1,}filteredBooks\s*=\s*books\.filter\(\(book\)\s*=>\s*book\.releaseYear\s*[<>=]\s*[0-9]\)")
+```
+
+You should call the `sort` higher order function by passing the `sortByYear` callback function on the `filteredBooks` array.
+
+```js
+assert.match(code, "(let|const)\s{1,}filteredBooksSorted\s*=\s*filteredBooks\.sort\(sortByYear\)")
+```
+
+You should call the `forEach` higher order function on the `filteredBooksSorted` array and print out the details of each book following the `<BOOK_TITLE> by <BOOK_AUTHOR> (<releaseYear>)` format.
+
+```js
+assert.match(code, "filteredBooksSorted\.forEach\(\(book\)\s*=>\s*\{\s*console\.log\(`\$\{book\.title\} by \$\{book\.authorName\} \(\$\{book\.releaseYear\}\)`\)\s*\}\)")
+```
```

Comment:
These should be updated based on my earlier comment 👍

Reply:
I have update the tests to check if the array is sorted in a different way. Working on the filtering test and the last one.